data_model/src/lib.rs

```diff
@@ -1804,6 +1810,89 @@ pub mod domain {
         }
     }
 
+    /// Represents path in IPFS. Performs some checks to ensure path validity.
+    ///
+    /// Should be constructed with `from_str()` method.
+    #[derive(
+        Debug,
+        Clone,
+        PartialEq,
+        Eq,
+        PartialOrd,
+        Ord,
+        Hash,
+        Decode,
+        Encode,
+        Deserialize,
+        Serialize,
+        IntoSchema,
+    )]
+    pub struct IpfsPath(String);
+
+    impl FromStr for IpfsPath {
+        type Err = ParseError;
+
+        fn from_str(string: &str) -> Result<Self, Self::Err> {
+            let mut subpath = string.split('/');
+            let empty = subpath.next().ok_or(ParseError {
+                reason: "Impossible error",
+            })?;
+
+            if empty.is_empty() {
+                let root_type = subpath.next().ok_or(ParseError {
+                    reason: "Expected root type, but nothing found",
+                })?;
+                let key = subpath.next().ok_or(ParseError {
+                    reason: "Expected at least one content id",
+                })?;
+
+                match root_type {
+                    "ipfs" | "ipld" => Self::check_cid(key)?,
+                    "ipns" => (),
+                    _ => {
+                        return Err(ParseError {
+                            reason: "Unexpected root type",
+                        })
+                    }
+                }
+            } else {
+                // by default if there is no prefix it's an ipfs or ipld path
+                Self::check_cid(empty)?;
+            }
+
+            for path in subpath {
+                Self::check_cid(path)?
+            }
+
+            Ok(IpfsPath(string.to_owned()))
+        }
+    }
+
+    impl AsRef<String> for IpfsPath {
+        fn as_ref(&self) -> &String {
+            &self.0
+        }
+    }
+
+    impl IpfsPath {
+        /// Instantly construct [`IpfsPath`] assuming the given `path` is valid.
+        #[inline]
+        pub fn test(path: String) -> Self {
+            Self(path)
+        }
+
+        /// Superficially checks IPFS `cid` (Content Identifier)
```

Comment:
`#[inline]`
One check and one conditional jump can be done using registers only. The reason is pre-allocated.